lib/msal-browser/src/cache/BrowserCacheManager.ts

```diff
@@ -145,200 +150,620 @@ export class BrowserCacheManager extends CacheManager {
+                oldSchemaData.secret,
+                base64Decode
+            );
+
+            const newIdTokenKey = this.generateCredentialKey(oldSchemaData);
+            const currentIdToken = this.getIdTokenCredential(
+                newIdTokenKey,
+                correlationId
+            );
+            const oldTokenHasSignInState =
+                Object.keys(claims).includes("signin_state");
+            const currentTokenHasSignInState =
+                currentIdToken &&
+                Object.keys(
+                    AuthToken.extractTokenClaims(
+                        currentIdToken.secret,
+                        base64Decode
+                    ) || {}
+                ).includes("signin_state");
+
+            /**
+             * Only migrate if:
+             * 1. Token doesn't yet exist in current schema
+             * 2. Old schema token has been updated more recently than the current one AND migrating it won't result in loss of KMSI state
+             */
             if (
-                this.cacheConfig.cacheLocation !==
-                    BrowserCacheLocation.LocalStorage ||
-                isEncrypted(parsedV0Value)
+                !currentIdToken ||
+                (oldSchemaData.lastUpdatedAt > currentIdToken.lastUpdatedAt &&
+                    (oldTokenHasSignInState || !currentTokenHasSignInState))
             ) {
-                const v1Key = `${CacheKeys.PREFIX}.${currentSchema}${CacheKeys.CACHE_KEY_SEPARATOR}${v0Key}`;
-                const rawV1Entry = this.browserStorage.getItem(v1Key);
-                if (!rawV1Entry) {
-                    upgradePromises.push(
-                        this.setUserData(
-                            v1Key,
-                            JSON.stringify(decryptedData),
-                            correlationId,
-                            parsedV0Value.lastUpdatedAt
-                        ).then(() => {
-                            v1Keys.push(v1Key);
-                            this.performanceClient.incrementFields(
-                                { upgradedCacheCount: 1 },
-                                correlationId
-                            );
-                        })
+                const tenantProfiles = account.tenantProfiles || [];
+                const tenantId =
+                    getTenantIdFromIdTokenClaims(claims) || account.realm;
+                if (
+                    tenantId &&
+                    !tenantProfiles.find((tenantProfile) => {
+                        return tenantProfile.tenantId === tenantId;
+                    })
+                ) {
+                    const newTenantProfile = buildTenantProfile(
+                        account.homeAccountId,
+                        account.localAccountId,
+                        tenantId,
+                        claims
+                    );
+                    tenantProfiles.push(newTenantProfile);
```

Comment:
Is the KMSI preference shared accross tenants? Or would running accross an ID token that doesn't have the kmsi claim result in that tenant profile being dropped in the migration/after the browser session ends?

Reply:
Yes, it's shared across tenants